### DISCUSSION TITLE
How to use raw sql with appwrite?

### DISCUSSION THREAD

**Discussion by usamashafiqs** [appwrite/appwrite#8047] [Q&A]
Is there is way to use raw SQL Query with appwrite?

**Comment by wellkilo** [comment]
Appwrite does not expose a raw SQL endpoint. The Databases/TablesDB APIs are the supported abstraction, and queries are built with Appwrite's Query helpers rather than arbitrary SQL strings. This is intentional: Appwrite enforces document/row permissions, schema validation, events, and database portability at the service layer; a raw SQL path would bypass those guarantees.

Use the closest supported option depending on the goal:

- Filtering/sorting/pagination: SDK Query methods such as equal, search, orderAsc/orderDesc, limit, cursorAfter.
- Aggregation or a complex server-side operation: create an Appwrite Function, validate the caller, then use the Server SDK to read/write through Appwrite APIs.
- Direct SQL against a self-hosted Appwrite internal database: technically possible with your infrastructure credentials, but unsupported and unsafe for application logic. Internal tables/migrations may change, and direct writes bypass Appwrite permissions/events and can corrupt its state.
- If arbitrary relational joins/SQL are a core product requirement, use a separate PostgreSQL/MySQL database for that data and access it from a Function/backend, while keeping Appwrite for Auth/Storage/etc.

So the short answer is: there is no supported raw-SQL API for Appwrite-managed data. If you share the particular SQL operation, it can usually be mapped to Query API calls or placed behind a Function. If this clarifies the boundary, please mark the reply as accepted.
